fix(AWS Deploy): Warn when invalid IAM policy to fetch lambda

trigaut opened the pull request:
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on a solution in the corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/test/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v10 is maintained. -->

<!--
⚠️⚠️ Ensure that the proposed change passes CI. Confirm that by running the following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/test/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide a link to the corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with a short description of made changes
-->

Closes: #8993

pgrzesik commented on lib/plugins/aws/deploy/lib/checkForChanges.js:
```diff
@@ -88,22 +88,37 @@ module.exports = {
         .then((res) => new Date(res.Configuration.LastModified))
-        .catch(() => new Date(0)); // Function is missing, needs to be deployed
+        .catch((err) => {
+          fetchError = err;
```
What do you think about inspecting the error here and setting e.g. `couldNotAccessFunction` flag instead of assigning the whole error?

pgrzesik commented on lib/plugins/aws/deploy/lib/checkForChanges.js:
```diff
@@ -88,22 +88,37 @@ module.exports = {
+          [
+            'WARNING: Iam policy misconfiguration: unable to get Lambda functions info.',
+            'Deployment will not be skipped even if service files did not change. ',
+            'Forcing deployment...',
```
I think we can remove `Forcing deployment...` part in the message, as it's already understood what will happen from the line above.

pgrzesik commented on lib/plugins/aws/deploy/lib/checkForChanges.js:
```diff
@@ -88,22 +88,37 @@ module.exports = {
+            'Forcing deployment...',
+            fetchError.message,
+          ].join('')
+        );
```
Let's also add the `Serverless` and `{color: 'orange'}` params to `this.serverless.cli.log` call similarly to: https://github.com/serverless/serverless/blob/b11171c70c8f5597e2644f7ea8ec82a17a9eee29/lib/plugins/aws/lib/checkIfEcrRepositoryExists.js#L19

trigaut replied:
When I put `{ color : 'orange' }` (I comitted the change) I obtain this. Do you know why ?

![serverlessWarning](https://user-images.githubusercontent.com/7863528/109791626-f107ae00-7c12-11eb-8b76-4f7978f8ca14.png)

trigaut replied:
Thanks for your suggestions ! I applied the changes

pgrzesik reviewed (changes requested):
Thanks @trigaut for this PR :raised_hands: I've added some minor comments/suggestions - please let me know what do you think. 

Additionally, it would be great to test that change and make sure it's emiting that warning as expected. Current tests live here: https://github.com/serverless/serverless/blob/b11171c70c8f5597e2644f7ea8ec82a17a9eee29/test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js#L683

trigaut commented:
> Thanks @trigaut for this PR  I've added some minor comments/suggestions - please let me know what do you think.
> 
> Additionally, it would be great to test that change and make sure it's emiting that warning as expected. Current tests live here:
> 
> https://github.com/serverless/serverless/blob/b11171c70c8f5597e2644f7ea8ec82a17a9eee29/test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js#L683

I will work on the test tomorrow

trigaut commented:
> Thanks @trigaut for this PR  I've added some minor comments/suggestions - please let me know what do you think.
> 
> Additionally, it would be great to test that change and make sure it's emiting that warning as expected. Current tests live here:
> 
> https://github.com/serverless/serverless/blob/b11171c70c8f5597e2644f7ea8ec82a17a9eee29/test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js#L683

I took your changes and added the test ;)

fredericbarthelet commented on test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:
```diff
@@ -1085,6 +1085,34 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
+      awsRequestStubMap: {
+        ...commonAwsSdkMock,
+        Lambda: {
+          getFunction: (() => {
+            const headObjectStub = sandbox.stub();
+            headObjectStub.throws({ providerError: { statusCode: 403 } });
+            return headObjectStub;
+          })(),
```
I think you can reduce  this  to :
```suggestion
          getFunction: sandbox.stub().throws({ providerError: { statusCode: 403 } }),
```

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:
```diff
@@ -1085,6 +1085,30 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
+    const { stdoutData } = await runServerless({
+      fixture: 'checkForChanges',
+      cliArgs: ['deploy'],
+      env: { AWS_CONTAINER_CREDENTIALS_FULL_URI: 'ignore' },
```
Is that ENV variable needed in this case?

trigaut replied:
No you are right I removed it thanks !

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:
```diff
@@ -1085,6 +1085,30 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
     expect(serverless.service.provider.shouldNotDeploy).to.equal(true);
   });
 
+  it.only('should print a warning if missing lambda:GetFunction permission', async () => {
```
I believe it's a leftover `only` here

trigaut replied:
:sweat:

pgrzesik reviewed (changes requested):
Looks great, thank you @trigaut :bow: I just have two very minor comments, after addressing them we're ready to merge :+1:

trigaut commented:
> Looks great, thank you @trigaut  I just have two very minor comments, after addressing them we're ready to merge

Thanks ! I comitted the changes you requested.

pgrzesik reviewed (approved):
Looks good, thank you @trigaut :bow:
